### PR TITLE
fix(eslint-plugin): [strict-boolean-expression] support falsy and truthy literals simultaneously

### LINKED ISSUE #6624: Bug: [strict-boolean-expression] using string literal union bypasses allowString: false
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.1-rc&sourceType=module&code=MYewdgzgLgBARgQwE4wLwwBQDMQgFwwBEChMAPkaRQK5gAmAplgJZgN0CUaAfDAN4AoGDGZZMAQhwgug4cNCQQAGwYA6JSADm2XByEwAvgINA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZfKsugIwHs+SAIZN6AD2LQUySnyaoMAbXDYciaND7RIKiAF0VAXxCGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const bar = (foo: "a" | "" | undefined) => {
  if (!foo) {
    console.log(foo)
  }
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": [
      "error"
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

i expect line 2 to error with
```
Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly.
```

### Actual Result

no error

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by kozlovvski**
## PR Checklist

- [x] Addresses an existing open issue: fixes #6624 
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Instead of checking if any of the string/number types is truthy to tell if combined type is truthy, we should do the opposite -> check if any of the types is falsy

### CODE REVIEW

File: packages/eslint-plugin/src/rules/strict-boolean-expressions.ts [hunk 1]
```diff
@@ -835,10 +835,12 @@ export default util.createRule<Options, MessageId>({
...
-        if (strings.some(type => type.isStringLiteral() && type.value !== '')) {
-          variantTypes.add('truthy string');
-        } else {
+        if (
+          strings.some(type => !type.isStringLiteral() || type.value === '')
+        ) {
           variantTypes.add('string');
+        } else {
+          variantTypes.add('truthy string');
         }
```
Comment (kozlovvski) [inline packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:844]: This also works, but is less efficient. 

```suggestion
        if (
          strings.every(type => type.isStringLiteral() && type.value !== '')
        ) {
          variantTypes.add('truthy string');
        } else {
          variantTypes.add('string');
        }
```

Comment (Josh-Cena) [reply packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:844]: @kozlovvski They both have the same efficiency :) They both bail out at the same expected number of iterations.

Comment (kozlovvski) [reply packages/eslint-plugin/src/rules/strict-boolean-expressions.ts:844]: @Josh-Cena didn't know that `every` returns early -> if this is the case, then probably the latter is more readable

**Review by Josh-Cena — COMMENTED** [review]
The fix looks great to me. Just a minor nit.

File: packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts [hunk 2]
```diff
@@ -965,6 +988,7 @@ if (y) {
         declare const x: number | null; if (x) {}
         (x?: number) => !x;
         <T extends number | null | undefined>(x: T) => x ? 1 : 0;
+        function foo(x: 0 | 1 | null) { if(!x) {} }
```
[the remark this answers is not in the document]
Comment (kozlovvski) [reply packages/eslint-plugin/tests/rules/strict-boolean-expressions.test.ts]: @Josh-Cena sure, it's fixed now 👍🏻

**Review by JoshuaKGoldberg — APPROVED** [review]
Agreed, looks great - thanks!

Kinda nifty how sometimes switching between `every` and `some` can fix bugs...

**Pull request merged**